File: swim-system-js/swim-core-js/@swim/structure/main/selector/LiteralSelector.ts
// Copyright 2015-2021 Swim inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {Murmur3, Numbers, Constructors} from "@swim/util";
import type {Output} from "@swim/codec";
import {Item} from "../Item";
import {Selector} from "./Selector";
import {AnyInterpreter, Interpreter} from "../"; // forward import

export class LiteralSelector extends Selector {
  constructor(item: Item, then: Selector) {
    super();
    Object.defineProperty(this, "item", {
      value: item.commit(),
      enumerable: true,
    });
    Object.defineProperty(this, "then", {
      value: then,
      enumerable: true,
    });
  }

  readonly item!: Item;

  override readonly then!: Selector;

  override forSelected<T>(interpreter: Interpreter,
                          callback: (interpreter: Interpreter) => T | undefined): T | undefined;
  override forSelected<T, S>(interpreter: Interpreter,
                             callback: (this: S, interpreter: Interpreter) => T | undefined,
                             thisArg: S): T | undefined;
  override forSelected<T, S>(interpreter: Interpreter,
                             callback: (this: S | undefined, interpreter: Interpreter) => T | undefined,
                             thisArg?: S): T | undefined {
    let selected: T | undefined;
    interpreter.willSelect(this);
    if (interpreter.scopeDepth !== 0) {
      const literal = this.item.evaluate(interpreter);
      if (literal.isDefined()) {
        // Push the literal onto the scope stack.
        interpreter.pushScope(literal);
        // Subselect the literal.
        selected = this.then.forSelected(interpreter, callback, thisArg);
        // Pop the literal off of the scope stack.
        interpreter.popScope();
      }
    }
    interpreter.didSelect(this, selected);
    return selected;
  }

  override mapSelected(interpreter: Interpreter,
                       transform: (interpreter: Interpreter) => Item): Item;
  override mapSelected<S>(interpreter: Interpreter,
                          transform: (this: S, interpreter: Interpreter) => Item,
                          thisArg: S): Item;
  override mapSelected<S>(interpreter: Interpreter,
                          transform: (this: S | undefined, interpreter: Interpreter) => Item,
                          thisArg?: S): Item {
    let result: Item;
    interpreter.willTransform(this);
    if (interpreter.scopeDepth !== 0) {
      let literal = this.item.evaluate(interpreter);
      if (literal.isDefined()) {
        // Push the literal onto the scope stack.
        interpreter.pushScope(literal);
        // Transform the literal.
        literal = this.then.mapSelected(interpreter, transform, thisArg);
        // Pop the literal off of the scope stack.
        interpreter.popScope();
      }
      result = literal;
    } else {
      result = Item.absent();
    }
    interpreter.didTransform(this, result);
    return result;
  }

  override substitute(interpreter: AnyInterpreter): Item {
    interpreter = Interpreter.fromAny(interpreter);
    const item = this.item.substitute(interpreter);
    let then = this.then.substitute(interpreter);
    if (!(then instanceof Selector)) {
      then = this.then;
    }
    return new LiteralSelector(item, then as Selector);
  }

  override andThen(then: Selector): Selector {
    return new LiteralSelector(this.item, this.then.andThen(then));
  }

  override get precedence(): number {
    return this.item.precedence;
  }

  override get typeOrder(): number {
    return 11;
  }

  override compareTo(that: unknown): number {
    if (that instanceof LiteralSelector) {
      let order = this.item.compareTo(that.item);
      if (order === 0) {
        order = this.then.compareTo(that.then);
      }
      return order;
    } else if (that instanceof Item) {
      return Numbers.compare(this.typeOrder, that.typeOrder);
    }
    return NaN;
  }

  override equivalentTo(that: unknown, epsilon?: number): boolean {
    if (this === that) {
      return true;
    } else if (that instanceof LiteralSelector) {
      return this.item.equivalentTo(that.item, epsilon)
          && this.then.equivalentTo(that.then, epsilon);
    }
    return false;
  }

  override equals(that: unknown): boolean {
    if (this === that) {
      return true;
    } else if (that instanceof LiteralSelector) {
      return this.item.equals(that.item) && this.then.equals(that.then);
    }
    return false;
  }

  override hashCode(): number {
    return Murmur3.mash(Murmur3.mix(Murmur3.mix(Constructors.hash(LiteralSelector),
        this.item.hashCode()), this.then.hashCode()));
  }

  override debug(output: Output): void {
    output = output.write("Selector").write(46/*'.'*/).write("literal").write(40/*'('*/)
        .debug(this.item).write(41/*')'*/);
    this.then.debugThen(output);
  }

  override debugThen(output: Output): void {
    // nop
  }

  override clone(): Selector {
    return new LiteralSelector(this.item.clone(), this.then.clone());
  }
}

File: swim-system-java/swim-core-java/swim.structure/src/main/java/swim/structure/selector/LiteralSelector.java
// Copyright 2015-2019 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.structure.selector;

import swim.codec.Output;
import swim.structure.Interpreter;
import swim.structure.Item;
import swim.structure.Selectee;
import swim.structure.Selector;
import swim.util.Murmur3;

/**
 * A {@link Selector} that, when {@link #evaluate evaluated}, evaluates and
 * yields {@code item} against {@code interpreter}.  This allows us to express
 * various selection criteria without having to implement the corresponding
 * dedicated {@code Selectors}.
 * <p>
 * Such behavior is <i>especially</i> useful in conjunction with {@link
 * FilterSelector FilterSelectors}. For example, to select {@code Items} in some
 * {@code Interpreter} such that the sum of the results of {@code selecting} "a"
 * and "b" is less than 5, we could do
 * <p>
 * {@code Selector.literal(Selector.get("a").plus(Selector.get("b")).lt(5))}
 * <p>
 * This, notably, does not require a "ConditionalSelector" and only needs
 * correct implementations of {@link swim.structure.operator.LtOperator} and
 * {@link swim.structure.operator.PlusOperator}.
 */
public final class LiteralSelector extends Selector {
  final Item item;
  final Selector then;

  public LiteralSelector(Item item, Selector then) {
    this.item = item.commit();
    this.then = then;
  }

  public Item item() {
    return this.item;
  }

  @Override
  public Selector then() {
    return this.then;
  }

  @Override
  public <T> T forSelected(Interpreter interpreter, Selectee<T> callback) {
    T selected = null;
    interpreter.willSelect(this);
    if (interpreter.scopeDepth() != 0) {
      final Item literal = this.item.evaluate(interpreter);
      if (literal.isDefined()) {
        // Push the literal onto the scope stack.
        interpreter.pushScope(literal);
        // Subselect the literal.
        selected = this.then.forSelected(interpreter, callback);
        // Pop the literal off of the scope stack.
        interpreter.popScope();
      }
    }
    interpreter.didSelect(this, selected);
    return selected;
  }

  @Override
  public Item mapSelected(Interpreter interpreter, Selectee<Item> transform) {
    final Item result;
    interpreter.willTransform(this);
    if (interpreter.scopeDepth() != 0) {
      Item literal = this.item.evaluate(interpreter);
      if (literal.isDefined()) {
        // Push the literal onto the scope stack.
        interpreter.pushScope(literal);
        // Transform the literal.
        literal = this.then.mapSelected(interpreter, transform);
        // Pop the literal off of the scope stack.
        interpreter.popScope();
      }
      result = literal;
    } else {
      result = Item.absent();
    }
    interpreter.didTransform(this, result);
    return result;
  }

  @Override
  public Item substitute(Interpreter interpreter) {
    final Item item = this.item.substitute(interpreter);
    Item then = this.then.substitute(interpreter);
    if (!(then instanceof Selector)) {
      then = this.then;
    }
    return new LiteralSelector(item, (Selector) then);
  }

  @Override
  public Selector andThen(Selector that) {
    return new LiteralSelector(this.item, this.then.andThen(then));
  }

  @Override
  public int precedence() {
    return this.item.precedence();
  }

  @Override
  public int typeOrder() {
    return 11;
  }

  @Override
  protected int compareTo(Selector that) {
    if (that instanceof LiteralSelector) {
      return compareTo((LiteralSelector) that);
    }
    return Integer.compare(typeOrder(), that.typeOrder());
  }

  int compareTo(LiteralSelector that) {
    int order = this.item.compareTo(that.item);
    if (order == 0) {
      order = this.then.compareTo(that.then);
    }
    return order;
  }

  @Override
  public boolean equals(Object other) {
    if (this == other) {
      return true;
    } else if (other instanceof LiteralSelector) {
      final LiteralSelector that = (LiteralSelector) other;
      return this.item.equals(that.item) && this.then.equals(that.then);
    }
    return false;
  }

  @Override
  public int hashCode() {
    if (hashSeed == 0) {
      hashSeed = Murmur3.seed(LiteralSelector.class);
    }
    return Murmur3.mash(Murmur3.mix(Murmur3.mix(hashSeed,
        this.item.hashCode()), this.then.hashCode()));
  }

  @Override
  public void debug(Output<?> output) {
    output = output.write("Selector").write('.').write("literal").write('(').debug(this.item).write(')');
    this.then.debugThen(output);
  }

  @Override
  public void debugThen(Output<?> output) {
    // nop
  }

  private static int hashSeed;
}
